quinn-proto/src/stream.rs

```diff
@@ -122,6 +122,46 @@ impl Recv {
         }
     }
 
+    pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, ReadError> {
+        assert!(
+            !self.unordered,
+            "cannot perform ordered reads following unordered reads on a stream"
+        );
+
+        for (data, offset) in self.buffered.drain(..) {
+            self.assembler.insert(offset, &data);
+        }
+
+        if !self.assembler.blocked() {
+            let n = self.assembler.read(buf);
+            // Only bother issuing stream credit if the peer wants to send more
+            if self.has_pending_incoming() {
+                self.max_data += n as u64;
+            }
+            Ok(n)
+        } else {
+            match self.state {
+                RecvState::ResetRecvd { error_code, .. } => {
+                    self.state = RecvState::Closed;
+                    Err(ReadError::Reset { error_code })
+                }
+                RecvState::Closed => unreachable!(),
+                RecvState::Recv { .. } => Err(ReadError::Blocked),
+                RecvState::DataRecvd { .. } => {
+                    self.state = RecvState::Closed;
+                    Err(ReadError::Finished)
+                }
+            }
+        }
+    }
+
+    pub fn has_pending_incoming(&self) -> bool {
```

Comment:
This naming is unclear to me. The condition we're actually checking here is whether the stream is in a state where the size (i.e. the final offset) is unknown, so maybe `has_unknown_size`, or invert it into `size_known`? Maybe even `pub fn size(&self) -> Option<u64>`, for use with `is_none()`?

Reply:
I renamed it to `receiving_size_known()`, inverting the condition. `size()` would've exposed more details than necessary, so I prefer the simpler variant that returns `bool`. I added `receiving` because it's not *just* about the value of the size, the actual state is important too.

Reply:
> it's not just about the value of the size, the actual state is important too.

I don't totally follow this. The current impl returns `true` for all states except recv with unknown size, including the non-receiving states like `ResetRecvd` and `Closed`.

Reply:
Huh, yeah, that was muddled. I've renamed it to `receiving_unknown_size()` now.